spec/core/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -124,12 +124,54 @@ interface Packet {
 
 Note that a `Packet` is never directly serialised. Rather it is an intermediary structure used in certain function calls that may need to be created or processed by modules calling the IBC handler.
 
+Question: why is this here? Isn't there a protobuf definition for Packet serialization which must be understood by all
```

Comment:
No, there is not a protobuf for a `Packet`, because it is never sent over the wire, the fields are implicitly committed to by Merkle path (store) keys.

Reply:
Thank you for the review. I will respond completely on Monday, but one question Chris. 

Please correct me if I am wrong here. In my understanding, when a Packet is passed to the ics-4 handler to send. That will:

- emit events that correspond to each field of the packet
  - one of those fields is app- specific data. The rest what I call "meta data" (defined by core protocol not app packet)
- The packet is serialized as protobuf and then hashed
- This hash is committed to in the Merkle store

In order to relay the packet, the relayer must parse the fields from the events and reform them into a protobuf message, which serialized to a preimage of the commitment on the source chain.

Is this correct? I am pretty sure there is a protobuf encoding used for the proofs/hashes and that is what I was referring to.

I want to do a similar thing with Acknowledgement as you do with Packet now. As close as I can do without breaking something. But mainly I need to understand better what the process is for Packet

Reply:
> The packet is serialized as protobuf and then hashed

No, this does not happen. The only thing that is stored is the commitment to the packet data, timeout height, and timeout timestamp. See [here](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#sending-packets) in the specification and [here](https://github.com/cosmos/ibc-go/blob/1ff0da6694d08e494a9c87d5b9737ab5312677a0/modules/core/04-channel/types/packet.go#L18) in the code. Perhaps we should consider changing this, but in any case we _do not_ need to serialise other fields (e.g. concerning channel source and destination) because they are implicitly committed to by the Merkle store keys.

Reply:
Thank you for that clarification. Very useful for me.

I feel that [hash function you linked](https://github.com/cosmos/ibc-go/blob/1ff0da6694d08e494a9c87d5b9737ab5312677a0/modules/core/04-channel/types/packet.go#L18):

`sha256(bigEndian(timeoutTimestamp) || bigEndian(timeoutRevNumber)  || bigEndian(timeoutRevHeight) || sha256(data))`

needs to be in the IBC spec, as it is required for a compatible implementation. At least it would clear a few points up for me.

It would also help me understand this better if the actual flow of commitments and reconstructing the proofs (eg. IBC from PoV of the relayer) was documented in another section. I feel this is critical use case and important for correctness/liveness arguments as well.